CSRF handling broken on master

### Are you looking for help?

when using `CSRFFilter` as a global filter and adding a CSRF field to a Form via ` @helper.CSRF.formField` Play will always fail with a 403 error and sometimes print a `RuntimeException` (when calling the Form via `GET` i.e. refreshing the browser with CTRL + R when the Form raised a 403):
```
Caused by: java.lang.RuntimeException: No CSRF token present!
	at scala.sys.package$.error(package.scala:27)
	at views.html.helper.CSRF$$anonfun$2.apply(CSRF.scala:29)
	at views.html.helper.CSRF$$anonfun$2.apply(CSRF.scala:29)
	at scala.Option.getOrElse(Option.scala:121)
	at views.html.helper.CSRF$.formField(CSRF.scala:29)
```

This does not happen on the 2.5.x branch.

However I still looking for more details and didn't had time lately to debug it more closely.
I guess the changes from Request Tags to the new Typed Request Maps introduced it (but it's just a guess).

### Play Version (2.5.x / etc)

master

### API (Scala / Java / Neither / Both)

Scala

### Operating System (Ubuntu 15.10 / MacOS 10.10 / Windows 10)

Mac 10.11?

### JDK (Oracle 1.8.0_72, OpenJDK 1.8.x, Azul Zing)

Java SE 1.8.0_112-b16

--- a/framework/src/play-filters-helpers/src/main/scala/play/filters/csrf/csrf.scala
+++ b/framework/src/play-filters-helpers/src/main/scala/play/filters/csrf/csrf.scala
@@ -211,9 +211,7 @@ object CSRF {
     /** Generate a token */
     def generateToken: String
     /** Compare two tokens */
-    def compareTokens(tokenA: String, tokenB: String): Boolean = {
-      java.security.MessageDigest.isEqual(tokenA.getBytes("utf-8"), tokenB.getBytes("utf-8"))
-    }
+    def compareTokens(tokenA: String, tokenB: String): Boolean
   }
 
   class TokenProviderProvider @Inject() (config: CSRFConfig, tokenSigner: CSRFTokenSigner) extends Provider[TokenProvider] {
@@ -226,14 +224,19 @@ object CSRF {
   class ConfigTokenProvider(config: => CSRFConfig, tokenSigner: CSRFTokenSigner) extends TokenProvider {
     lazy val underlying = new TokenProviderProvider(config, tokenSigner).get
     def generateToken = underlying.generateToken
+    override def compareTokens(tokenA: String, tokenB: String) = underlying.compareTokens(tokenA, tokenB)
   }
 
   class SignedTokenProvider(tokenSigner: CSRFTokenSigner) extends TokenProvider {
     def generateToken = tokenSigner.generateSignedToken
+    def compareTokens(tokenA: String, tokenB: String) = tokenSigner.compareSignedTokens(tokenA, tokenB)
   }
 
   class UnsignedTokenProvider(tokenSigner: CSRFTokenSigner) extends TokenProvider {
     def generateToken = tokenSigner.generateToken
+    override def compareTokens(tokenA: String, tokenB: String) = {
+      java.security.MessageDigest.isEqual(tokenA.getBytes("utf-8"), tokenB.getBytes("utf-8"))
+    }
   }
 
   /**

--- a/framework/src/play-filters-helpers/src/test/scala/play/filters/csrf/CSRFCommonSpecs.scala
+++ b/framework/src/play-filters-helpers/src/test/scala/play/filters/csrf/CSRFCommonSpecs.scala
@@ -6,12 +6,13 @@ package play.filters.csrf
 import org.specs2.matcher.MatchResult
 import org.specs2.mutable.Specification
 import play.api.Application
-import play.api.http.{ ContentTypeOf, ContentTypes, SecretConfiguration, Writeable }
+import play.api.http.{ ContentTypeOf, ContentTypes, SecretConfiguration }
 import play.api.inject.guice.GuiceApplicationBuilder
 import play.api.libs.crypto._
 import play.api.libs.ws._
 import play.api.mvc.{ Handler, Session }
 import play.api.test.{ PlaySpecification, TestServer }
+import play.filters.csrf.CSRF.{ SignedTokenProvider, UnsignedTokenProvider }
 
 import scala.concurrent.Future
 import scala.reflect.ClassTag
@@ -27,7 +28,9 @@ trait CSRFCommonSpecs extends Specification with PlaySpecification {
 
   def inject[T: ClassTag](implicit app: Application) = app.injector.instanceOf[T]
 
-  val crypto = new DefaultCSRFTokenSigner(new DefaultCookieSigner(SecretConfiguration(CRYPTO_SECRET)), java.time.Clock.systemUTC())
+  val tokenSigner = new DefaultCSRFTokenSigner(new DefaultCookieSigner(SecretConfiguration(CRYPTO_SECRET)), java.time.Clock.systemUTC())
+  val signedTokenProvider = new SignedTokenProvider(tokenSigner)
+  val unsignedTokenProvider = new UnsignedTokenProvider(tokenSigner)
 
   val Boundary = "83ff53821b7c"
   def multiPartFormDataBody(tokenName: String, tokenValue: String) = {
@@ -132,13 +135,13 @@ trait CSRFCommonSpecs extends Specification with PlaySpecification {
     "work with signed session tokens" in {
       def csrfCheckRequest = buildCsrfCheckRequest(sendUnauthorizedResult = false)
       def csrfAddToken = buildCsrfAddToken()
-      def generate = crypto.generateSignedToken
+      def generate = signedTokenProvider.generateToken
       def addToken(req: WSRequest, token: String) = req.withSession(TokenName -> token)
       def getToken(response: WSResponse) = {
         val session = response.cookies.find(_.name.exists(_ == Session.COOKIE_NAME)).flatMap(_.value).map(Session.decode)
         session.flatMap(_.get(TokenName))
       }
-      def compareTokens(a: String, b: String) = crypto.compareSignedTokens(a, b) must beTrue
+      def compareTokens(a: String, b: String) = signedTokenProvider.compareTokens(a, b) must beTrue
 
       sharedTests(csrfCheckRequest, csrfAddToken, generate, addToken, getToken, compareTokens, FORBIDDEN)
 
@@ -163,7 +166,7 @@ trait CSRFCommonSpecs extends Specification with PlaySpecification {
         csrfAddToken(req => addToken(req, token).get()) { response =>
           // it shouldn't be equal, to protect against BREACH vulnerability
           response.body must_!= token
-          crypto.compareSignedTokens(token, response.body) must beTrue
+          signedTokenProvider.compareTokens(token, response.body) must beTrue
         }
       }
     }
@@ -171,7 +174,7 @@ trait CSRFCommonSpecs extends Specification with PlaySpecification {
     "work with unsigned session tokens" in {
       def csrfCheckRequest = buildCsrfCheckRequest(false, "play.filters.csrf.token.sign" -> "false")
       def csrfAddToken = buildCsrfAddToken("play.filters.csrf.token.sign" -> "false")
-      def generate = crypto.generateToken
+      def generate = unsignedTokenProvider.generateToken
       def addToken(req: WSRequest, token: String) = req.withSession(TokenName -> token)
       def getToken(response: WSResponse) = {
         val session = response.cookies.find(_.name.exists(_ == Session.COOKIE_NAME)).flatMap(_.value).map(Session.decode)
@@ -185,10 +188,10 @@ trait CSRFCommonSpecs extends Specification with PlaySpecification {
     "work with signed cookie tokens" in {
       def csrfCheckRequest = buildCsrfCheckRequest(false, "play.filters.csrf.cookie.name" -> "csrf")
       def csrfAddToken = buildCsrfAddToken("play.filters.csrf.cookie.name" -> "csrf")
-      def generate = crypto.generateSignedToken
+      def generate = signedTokenProvider.generateToken
       def addToken(req: WSRequest, token: String) = req.withCookies("csrf" -> token)
       def getToken(response: WSResponse) = response.cookies.find(_.name.exists(_ == "csrf")).flatMap(_.value)
-      def compareTokens(a: String, b: String) = crypto.compareSignedTokens(a, b) must beTrue
+      def compareTokens(a: String, b: String) = signedTokenProvider.compareTokens(a, b) must beTrue
 
       sharedTests(csrfCheckRequest, csrfAddToken, generate, addToken, getToken, compareTokens, FORBIDDEN)
     }
@@ -196,7 +199,7 @@ trait CSRFCommonSpecs extends Specification with PlaySpecification {
     "work with unsigned cookie tokens" in {
       def csrfCheckRequest = buildCsrfCheckRequest(false, "play.filters.csrf.cookie.name" -> "csrf", "play.filters.csrf.token.sign" -> "false")
       def csrfAddToken = buildCsrfAddToken("play.filters.csrf.cookie.name" -> "csrf", "play.filters.csrf.token.sign" -> "false")
-      def generate = crypto.generateToken
+      def generate = unsignedTokenProvider.generateToken
       def addToken(req: WSRequest, token: String) = req.withCookies("csrf" -> token)
       def getToken(response: WSResponse) = response.cookies.find(_.name.exists(_ == "csrf")).flatMap(_.value)
       def compareTokens(a: String, b: String) = a must_== b
@@ -207,7 +210,7 @@ trait CSRFCommonSpecs extends Specification with PlaySpecification {
     "work with secure cookie tokens" in {
       def csrfCheckRequest = buildCsrfCheckRequest(false, "play.filters.csrf.cookie.name" -> "csrf", "play.filters.csrf.cookie.secure" -> "true")
       def csrfAddToken = buildCsrfAddToken("play.filters.csrf.cookie.name" -> "csrf", "play.filters.csrf.cookie.secure" -> "true")
-      def generate = crypto.generateSignedToken
+      def generate = signedTokenProvider.generateToken
       def addToken(req: WSRequest, token: String) = req.withCookies("csrf" -> token)
       def getToken(response: WSResponse) = {
         response.cookies.find(_.name.exists(_ == "csrf")).flatMap { cookie =>
@@ -215,7 +218,7 @@ trait CSRFCommonSpecs extends Specification with PlaySpecification {
           cookie.value
         }
       }
-      def compareTokens(a: String, b: String) = crypto.compareSignedTokens(a, b) must beTrue
+      def compareTokens(a: String, b: String) = signedTokenProvider.compareTokens(a, b) must beTrue
 
       sharedTests(csrfCheckRequest, csrfAddToken, generate, addToken, getToken, compareTokens, FORBIDDEN)
     }
@@ -223,10 +226,10 @@ trait CSRFCommonSpecs extends Specification with PlaySpecification {
     "work with checking failed result" in {
       def csrfCheckRequest = buildCsrfCheckRequest(true, "play.filters.csrf.cookie.name" -> "csrf")
       def csrfAddToken = buildCsrfAddToken("play.filters.csrf.cookie.name" -> "csrf")
-      def generate = crypto.generateSignedToken
+      def generate = signedTokenProvider.generateToken
       def addToken(req: WSRequest, token: String) = req.withCookies("csrf" -> token)
       def getToken(response: WSResponse) = response.cookies.find(_.name.exists(_ == "csrf")).flatMap(_.value)
-      def compareTokens(a: String, b: String) = crypto.compareSignedTokens(a, b) must beTrue
+      def compareTokens(a: String, b: String) = signedTokenProvider.compareTokens(a, b) must beTrue
 
       sharedTests(csrfCheckRequest, csrfAddToken, generate, addToken, getToken, compareTokens, UNAUTHORIZED)
     }

--- a/framework/src/play-filters-helpers/src/test/scala/play/filters/csrf/CSRFFilterSpec.scala
+++ b/framework/src/play-filters-helpers/src/test/scala/play/filters/csrf/CSRFFilterSpec.scala
@@ -81,7 +81,7 @@ class CSRFFilterSpec extends CSRFCommonSpecs {
               .map(Results.Ok(_))
               .getOrElse(Results.NotFound))
       }){ ws =>
-        val token = crypto.generateSignedToken
+        val token = signedTokenProvider.generateToken
         await(ws.url("http://localhost:" + testServerPort).withSession(TokenName -> token)
           .post(Map("foo" -> "bar", TokenName -> token))).body must_== "bar"
       }
@@ -112,7 +112,7 @@ class CSRFFilterSpec extends CSRFCommonSpecs {
       .build()
 
     "feed a not fully buffered body once a check has been done and passes" in new WithServer(notBufferedFakeApp, testServerPort) {
-      val token = crypto.generateSignedToken
+      val token = signedTokenProvider.generateToken
       val ws = inject[WSClient]
       val response = await(ws.url("http://localhost:" + port).withSession(TokenName -> token)
         .withHeaders(CONTENT_TYPE -> "application/x-www-form-urlencoded")
@@ -134,10 +134,10 @@ class CSRFFilterSpec extends CSRFCommonSpecs {
     "work with a Java error handler" in {
       def csrfCheckRequest = buildCsrfCheckRequestWithJavaHandler()
       def csrfAddToken = buildCsrfAddToken("csrf.cookie.name" -> "csrf")
-      def generate = crypto.generateSignedToken
+      def generate = signedTokenProvider.generateToken
       def addToken(req: WSRequest, token: String) = req.withCookies("csrf" -> token)
       def getToken(response: WSResponse) = response.cookies.find(_.name.exists(_ == "csrf")).flatMap(_.value)
-      def compareTokens(a: String, b: String) = crypto.compareSignedTokens(a, b) must beTrue
+      def compareTokens(a: String, b: String) = signedTokenProvider.compareTokens(a, b) must beTrue
 
       sharedTests(csrfCheckRequest, csrfAddToken, generate, addToken, getToken, compareTokens, UNAUTHORIZED)
     }

--- a/framework/src/play-filters-helpers/src/test/scala/play/filters/csrf/JavaCSRFActionSpec.scala
+++ b/framework/src/play-filters-helpers/src/test/scala/play/filters/csrf/JavaCSRFActionSpec.scala
@@ -82,9 +82,9 @@ class JavaCSRFActionSpec extends CSRFCommonSpecs {
     )) { implicit app =>
       { case _ => javaAction[JavaCSRFActionSpec.MyAction]("getToken", myAction.getToken) }
     } { ws =>
-      lazy val token = crypto.generateSignedToken
+      lazy val token = signedTokenProvider.generateToken
       val returned = await(ws.url("http://localhost:" + testServerPort).withSession(TokenName -> token).get()).body
-      crypto.compareSignedTokens(token, returned) must beTrue
+      signedTokenProvider.compareTokens(token, returned) must beTrue
     }
   }
 

